File: github_cpp/14/59.cpp
#include "RadixSort.h"



int RadixSort::getMax(int *array, int n) {
    int mx = array[0];
    for (int i = 1; i < n; i++)
        if (array[i] > mx)
            mx = array[i];
    return mx;
}


void RadixSort::countSort(int *array, int n, int exp) {
    int output[n]; 
    int i, count[10] = {0};

    
    for (i = 0; i < n; i++) {
        count[(array[i] / exp) % 10]++;
    }

    
    
    for (i = 1; i < 10; i++) {
        count[i] += count[i - 1];
    }

    
    for (i = n - 1; i >= 0; i--) {
        output[count[(array[i] / exp) % 10] - 1] = array[i];
        count[(array[i] / exp) % 10]--;
    }

    
    
    for (i = 0; i < n; i++) {
        array[i] = output[i];
    }
}


void RadixSort::sort(int *array, int n) const {
    
    int m = getMax(array, n);

    
    
    
    for (int exp = 1; m / exp > 0; exp *= 10) {
        countSort(array, n, exp);
    }
}

File: github_code_index/java/11/59.java
package sortingAlgorithms.algorithms.linear;

import sortingAlgorithms.algorithms.SortingAlgorithm;

import java.util.ArrayList;
import java.util.List;

/**
 * Radix sort is a non-comparative integer sorting algorithm that sorts data with integer keys by
 * grouping keys by the individual digits which share the same significant position and value. A
 * positional notation is required, but because integers can represent strings of characters (e.g.,
 * names or dates) and specially formatted floating point numbers, radix sort is not limited to
 * integers. [http://en.wikipedia.org/wiki/Radix_sort]
 */
public class RadixSort implements SortingAlgorithm {

    /**
     * @param list - contain array with unsorted integer values.
     * @throws IllegalArgumentException
     * @return List<Integer>
     */
    @Override
    @SuppressWarnings("unchecked")
    public List<Integer> sort(List<Integer> list) {

        // checking input parameter for null
        if (list == null) {
            throw new IllegalArgumentException("ArrayList not specified!");
        }

        // To get a digit we can first divide number with 10^i and then get the remainder
        // after dividing by 10, where i = 0 to 2*Number of digits in N;
        int exp = 1;

        // the number of digits in N^2 can be maximum double of the digits in N
        int maxDigits = 2;

        for (int i = 0; i < maxDigits; i++) {

            // initiating multidimensional list of lists of size 10
            List[] bucketList = new ArrayList[10];
            for (int k = 0; k < 10; k++) {
                bucketList[k] = new ArrayList<Integer>();
            }
            for (Integer aList : list) {
                int number = (aList / exp) % 10;
                bucketList[number].add(aList); // producing warning: unchecked assignment - suppressed.
            }
            exp *= 10;
            int index = 0;
            for (int k = 0; k < 10; k++) {
                for (Object num : bucketList[k]) {
                    list.set(index, (Integer) num);
                    index++;
                }
            }
        }
        return list;
    }
}
